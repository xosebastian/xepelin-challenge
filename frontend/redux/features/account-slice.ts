import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'

export interface AccountState {
  id: string,
  balance: number,
}

const initialState: AccountState = {
  id: '',
  balance: 0,
}

export const accountSlice = createSlice({
  name: 'account',
  initialState,
  reducers: {
    create: (state, action: PayloadAction<{id : string}>) => {
      state.id = action.payload.id
    },
    clear: () => {
      return initialState
    },
    update: (state, action: PayloadAction<{balance: number}>) => {
      state.balance = action.payload.balance
    },
    transactions: (state, action: PayloadAction<{balance: number, type: string}>) => {
      if (action.payload.type === 'DEPOSIT') {
        state.balance += action.payload.balance
      }else if (action.payload.type === 'WITHDRAWAL') {
        state.balance -= action.payload.balance
      }
    },
  },
})


export const { create, clear, update, transactions } = accountSlice.actions

export default accountSlice.reducer