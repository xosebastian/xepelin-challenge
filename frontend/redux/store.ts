import { configureStore } from '@reduxjs/toolkit'
import accountReducer from './features/account-slice'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'

export const store = configureStore({
  reducer: {
    account: accountReducer,
  },
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;