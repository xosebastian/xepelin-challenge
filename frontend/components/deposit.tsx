"use client";
import { useAppSelector } from "@/redux/store";
import { DepositForm } from "./deposit-form";
export function Deposit () {


  const accountId = useAppSelector((state) => state.account.id);

  if(!accountId) {
    return (
      <div className="flex flex-col space-y-1.5 p-6">
        <h3 className="text-2xl font-semibold leading-none tracking-tight">
          Desposit to account
        </h3>
      </div>
    );
  }


  return <DepositForm accountId={accountId} />;  
}