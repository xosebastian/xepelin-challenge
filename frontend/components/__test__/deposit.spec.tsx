import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";

import { store } from "@/redux/store";
import { Deposit } from "../deposit";
import { create } from "@/redux/features/account-slice";

describe("Deposit", () => {
  beforeEach(() => {
    store.dispatch({
      type: "account/reset",
    });
  });

  it("should render a message when accountId is not defined", () => {
    const { container } = render(
      <Provider store={store}>
        <Deposit />
      </Provider>
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByText("Desposit to account")).toBeDefined();
  });

  it("should render the DepositForm when accountId is defined", () => {
    store.dispatch(
      create({
        id: "user-id",
      })
    );

    const { container } = render(
      <Provider store={store}>
        <Deposit />
      </Provider>
    );

    expect(container).toMatchSnapshot();
    expect(screen.queryByText("Desposit to account")).toBeNull();
  });
});
