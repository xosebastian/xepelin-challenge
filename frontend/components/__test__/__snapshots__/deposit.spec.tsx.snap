// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deposit should render a message when accountId is not defined 1`] = `
<div>
  <div
    class="flex flex-col space-y-1.5 p-6"
  >
    <h3
      class="text-2xl font-semibold leading-none tracking-tight"
    >
      Desposit to account
    </h3>
  </div>
</div>
`;

exports[`Deposit should render the DepositForm when accountId is defined 1`] = `
<div>
  <div
    class="flex flex-col space-y-1.5 p-6"
  >
    <h3
      class="text-2xl font-semibold leading-none tracking-tight"
    >
      Deposit to account
    </h3>
  </div>
  <form
    class="space-y-2 pl-6"
  >
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        Balance
      </label>
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r0:-form-item"
        name="balance"
        type="number"
        value="0"
      />
    </div>
    <button
      class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
      type="submit"
    >
      Deposit
    </button>
  </form>
</div>
`;
